### ISSUE TITLE
ECMAScript

### ISSUE THREAD

**Issue by birkskyum** [maplibre/maplibre-gl-js#1595]
ESM is the new standard with import/export etc. We don't use it now because there are some old mapbox dependencies that only export .cjs, but nothing more than we can handle. Also, firefox has for 5 years struggled to support this in web workers, and now it's about time to move on:

### Firefox
Lack ESM support for web workers: https://caniuse.com/mdn-api_worker_worker_ecmascript_modules
5+ years of ongoing effort: https://bugzilla.mozilla.org/show_bug.cgi?id=1247687

### Benchmarks
I once attempted to move the benchmark to ESM #964 , but found the change wasn't backward compatible with the already uploaded benchmarks of prev versions, so it was rolled back #969 . 

This can be a tracking issue where we first map the dependencies we have that aren't compatible, and then keep this updated when they are updated

**Comment by HarelM** [comment]
Do you mean that we currently export maplibre as commonjs and we should export it as esm?
I think the problem of switching out of commonjs besides the FF support was the ability to have the worker URL pushed while building this.
There are some complicated build steps to facilitate for this which I think are very hard to split.
I've looked at this when I tried to understand if we could leverage tree shaking here: #977
I generally think this is a good idea, even better if we can introduce it as part of 3.0 version as a breaking change.
But I'm not sure how to solve all the build complexity with the workers url and also the shared code between the main thread and the worker thread...

**Comment by birkskyum** [comment]
Sure, I think we might be able to take small steps towards it. A major blocker previously was i.e. also all this node assert that would break the dev build if exported as esm. The first step is likely to have the build as esm.

**Comment by backspaces** [comment]
The published dist/ dir already has multiple versions: https://unpkg.com/browse/maplibre-gl@2.4.0/dist/
And indeed, the individual src files appear use import/export, although "bare".

So shouldn't it be easy to just add a new one: maplibre-gl-esm.js?
And add a "module": "dist/maplibre-gl-esm.js" to package.json?
That seems to be the common approach for folks wanting to support both forms.

**Comment by geraldo** [comment]
> import maplibreGl from 'https://cdn.skypack.dev/maplibre-gl'

@backspaces That doesn't work for me:

`var map = new maplibreGl.Map()`

gives me an `[ERR_UNSUPPORTED_ESM_URL_SCHEME]` error.

**Comment by backspaces** [comment]
Try it using this:

```
<!DOCTYPE html>
<html>
<head>
    <title>Imports</title>
</head>
<body>
    <script type="module">
        import maplibregl from 'https://cdn.skypack.dev/maplibre-gl'
        console.log('https://cdn.skypack.dev/maplibre-gl', maplibregl)
    </script>
</body>
</html>
```

**Comment by geraldo** [comment]
Thanks, that works fine when run directly in the browser, and even shows a basic map:

```
var map = new maplibregl.Map({
	container: 'map',
	style: 'https://demotiles.maplibre.org/style.json',
	center: [0, 0],
	zoom: 0
})
```

But when I try to run it with node (using [this https module loader](https://stackoverflow.com/a/66921779/1996896)), `new maplibregl.Map` does throw an error:

> ReferenceError: devicePixelRatio is not defined

**Comment by birkskyum** [comment]
The devicePixelRatio likely fail because there is no device/display. What do you hope to achieve running it in node ? Depending on your use-case, you might want to use the node distribution of maplibre-gl-native instead https://www.npmjs.com/package/@maplibre/maplibre-gl-native

**Comment by geraldo** [comment]
The use case would be including Maplibre as dependency, for example offering a [maplibre renderer in Maputnik](https://github.com/maputnik/editor/issues/781) or other kind of web apps depending on Maplibre. So it still would be web and not Android nor iOS, that's why I think using maplibre-gl-native would not work right?

**Comment by birkskyum** [comment]
@acalcutt , what is your take on this?

**Comment by birkskyum** [comment]
Related to https://github.com/maplibre/maplibre/discussions/159 through tree-shaking

**Comment by HarelM** [comment]
I have made a lot of change in version 4 and basically removed the global object that existed before.
It might be interesting to try out again to export the library in both esm and common js.

**Comment by HarelM** [comment]
The article which I'm looking at right now is the following:
https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html

Since we add a package.json file to dist folder in order to force bundling tools to treat maplibre as common js I think we'll need to place the esm output in a folder called mjs or something to avoid it being treated as commonjs.

I'll do some experiments locally and see if I can get to a working version or not.
Since this is not a replacement and only an addition there's no breaking change (hopefully) and I'll be able to push it to a regular version and not a breaking change version. later on we can decide if and when to deprecated commonjs build output.

**Comment by geraldo** [comment]
Seems to work now as recent Maputnik versions (v2.0) use [MapLibre 4.1.2 as dependency](https://github.com/maplibre/maputnik#develop).

**Comment by HarelM** [comment]
Below is a link to an article on esm with rollup and worker configuration:
https://justinribeiro.com/chronicle/2020/07/17/building-module-web-workers-for-cross-browser-compatibility-with-rollup/
This is to overcome the amd usage for the code to load it to the worker.
The above suggest to export the worker code as text - which means no reuse of shared code between the worker and the main thread.
If there's no code sharing then the csp approach is probably better.
Seems like esm and loading the worker code automatically are contradicting one another.
So the current build configuration is the best solution as far as I understand from bundle size perspective.
We might consider introducing esm to the csp build, but this will have very small impact as I believe most users are not using the csp build.

If anyone has more information on this or knows this topic more than I do please feel free to let me know.

**Comment by HarelM** [comment]
@timocov sorry to tag and feel free to unsubscribe, but it would be great if you can take a peak at this and let me know if I'm missing anything - I know you have a lot of knowledge/experience around rollup, build, esm, commonjs amd etc, and any input would be helpful here.

**Comment by unphased** [comment]
Can I get a bit of clarification on this, I see errors like this when I run `tsc`:

```
node_modules/maplibre-gl/dist/maplibre-gl.d.ts:4:21 - error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("@mapbox/tiny-sdf")' call instead.
```

What does this mean? I need to treat Maplibre GL as a CJS dependency?

**Comment by HarelM** [comment]
Yes, currently you should treat maplibre as commonjs until this issue is resolved.

**Comment by unphased** [comment]
For what it's worth... I did spend a day and a half or so fighting tsconfig and such with my neovim dev environment, but I do actually have maplibregl typescript types getting fetched properly by tsserver in a ESM project. Here's my current config:

```
{
  "exclude": [ "node_modules" ],
  "include": [ "frontend/**/*.ts" ],
  "compilerOptions": {
    "allowJs": true,                         
    "module": "NodeNext",
    "target": "ESNext",
    "lib": [ "ESNext", "dom" ],         
    "outDir": "./build",
    "moduleResolution": "nodenext",
    "rootDir": "./",
}
```

I simply have the types package set up (`"@types/maplibre-gl": "^1.14.0",`), perhaps that's the more important thing.

**Comment by wmertens** [comment]
Am I understanding correctly that most of the code size is inside the worker? And that all of the code is used?

In that case ESM builds wouldn't make much of a difference.

It would be different if lots of worker code is not used, in that case it would be better to use dynamic imports and a cjs build for it, but still not ESM.

**Comment by HarelM** [comment]
There's a lot (I think) of shared code between the workers and the main thread, which is what the current build try to improve in terms of package size - i.e. pack the shared code once to be used by both the worker and the main thread without increasing the package size.

**Comment by birkskyum** [comment]
@HarelM , I think the situation has developed quite a lot since the article you linked [here](https://github.com/maplibre/maplibre-gl-js/issues/1595#issuecomment-2106196032) was written in 2020. As it write a few sections down "All things equal if we ran this in Chrome we’d be golden and we’ve have a magical day.". Thing is, both safari (2021) and firefox (2023) have stabilized support for esm modules in workers since, so now we're close to a point where esm can simply be used throughout.

The current support levels for modules in workers is [tracked here](https://caniuse.com/?search=worker%20module) - I'm not sure how high it needs to be, but it's getting there. ESM won't make it into v5 anyways, adding at least half a year, and we have blockers still (#4735), so when we're technically ready, we might re-assess the esm support level and find it's sufficient - and save the majority of trouble outlined in the bundling article.

**Comment by HarelM** [comment]
@birkskyum is there a progress in the area of sharing code between worker thread and main thread with esm? I can't find anything on that...
If so, can you link to an article that shows how to achieve it?

**Comment by birkskyum** [comment]
Progress? I'm not familiar with any ongoing work on browser apis in that direction.

I don't think the ESM modules will make code sharing between worker/main better or worse - is there any problem with the current rollup bundle split strategy of client/worker/shared? and is that in any way tied to the esm format?

**Comment by wmertens** [comment]
@birkskyum I wonder if bundle size would be smaller with ESM if you're not using summery features?

**Comment by HarelM** [comment]
> is there any problem with the current rollup bundle split strategy of client/worker/shared? and how does that relate to the module format?

It's not that there is a problem, but as far as I understand, the code sharing that is performed right now can't be achieved with esm.
This might have changed since I wrote the comment above, so I asked if there was a progress in this aspect in respect to esm.

**Comment by backspaces** [comment]
I have switched from cdn.skypack.dev to  esm.sh:
```
    import maplibregl from 'https://esm.sh/maplibre-gl@3.4.0'
```
Do we know what the maplibre team plans to do to support es6 modules?

**Comment by smnbbrv** [comment]
> A different approach can be to ship ESM main thread, worker, and shared code as part of the package and add a section in the readme to guide people how to use the ESM version, how to set it up, much like the CSP version (or only do ESM for the CSP version).

This sounds like a way to go. I don't see the need in shared logic here, because it should anyway be loaded within the worker in a parallel thread. See how [monaco editor](https://github.com/microsoft/monaco-editor/blob/main/docs/integrate-esm.md) solves the very same problem

> I'm not sure it will solve anything, as I believe most people would still use the "easy" distribution, and I'm not sure you'll actually be able to tree-shake anything with the ESM version.

Currently people do not have any option and we cannot know whether they will or won't use it. Wordpress users of course would choose a single bundle, the bundler users won't. Regarding the tree-shaker, as long as there is no ESM output, there is nothing we can say how great the impact would be. 

Having 1MB js file that is loaded in main thread is just a killer for any public website. Currently the only solution is to embed the map on the page bottom lazily with a hope user won't scroll there; that's the only way to deal with website performance. If at least half of it would be moved to worker, it's a great win since worker is loaded, parsed and processed in a separate thread.

**Comment by HarelM** [comment]
A CSP version of maplibre-gl-js is shipped and can be used, that's not accurate that there's currently no solution.
Moreover, I've just went through the Monaco editor link you added on how to install it on different bundlers and I'm pretty sure no one from the community is planning on maintaining these plugins and wrappers judging from how old this thread is, also the complexity from the docs is overwhelming, even from a veteran JS developer as myself.
@smnbbrv if this is something you would like to push forward as it seems this is a pain point for your use case a PR would be very much welcomed.

**Comment by birkskyum** [comment]
Here's an overview of module types of our deps

<img width="1298" alt="Image" src="https://github.com/user-attachments/assets/90e0f0f7-50ff-4cc0-81e6-bd41947ada10" />

**Comment by birkskyum** [comment]
Updated stats 
<img width="1041" height="746" alt="Image" src="https://github.com/user-attachments/assets/cea72f79-f4d7-46a7-a057-c39e2d92d987" />

@mwilsnd ,  i wonder why the new vt-pbf doesn't show up as esm, given that the [rollup output format](https://github.com/maplibre/vt-pbf/blob/main/rollup.config.ts#L12) is set to 'es', and the tsconfig.json has `"module": "esnext"`. Maybe a `"type": "module"` in the package.json is the missing piece

**Comment by mwilsnd** [comment]
Looks like that's probably why, I'll update it.

**Comment by WAKayser** [comment]
I was wondering why global-prefix was part of the dependencies. It was once added for stylelint I guess, but its worth checking if that is still needed. https://github.com/maplibre/maplibre-gl-js/pull/1485#discussion_r945254545

**Comment by birkskyum** [comment]
@WAKayser , i checked just now, and we can safely remove it.

**Comment by birkskyum** [comment]
This is all that's left for dependencies
<img width="1266" height="652" alt="Image" src="https://github.com/user-attachments/assets/3410d630-c051-4cc8-9eaa-a978ea18e944" />

**Comment by HarelM** [comment]
In continue to the meeting today about ESM, here are some links related to what I talked about which highlight the history and complexity of what is being made here:
- https://github.com/rollup/rollup/issues/1029
- https://github.com/mapbox/mapbox-gl-js/issues/5939

**Comment by birkskyum** [comment]
@HarelM , the solution implemented in pr below is running the map just fine with only esm. The module just have to be imported slightly different, as demonstrated by the update example file.
- #6254

**Comment by HarelM** [comment]
There is an assumption there on where the worker code is, which might not work in some cases, especially with bundles. 
Also I would recommend adding it instead of creating a breaking change. 
I never said it wasn't possible, the problem I faced was around the worker code and shared code, with regards to bundle size, which I'm interesting to see if it increases or not.

**Comment by birkskyum** [comment]
So far bundle size looks good - if anything, there seems to be a 28kb reduction in prod bundle size of maplibre-gl.js -> .mjs

<img width="1456" height="732" alt="Image" src="https://github.com/user-attachments/assets/22c215b2-2b66-472b-ac95-3e22e2f9b780" />

unmodified main branch

<img width="1466" height="410" alt="Image" src="https://github.com/user-attachments/assets/c347c6fc-9f68-4b25-9191-51a7cd7da7ed" />

**Comment by birkskyum** [comment]
> I would recommend adding it instead of creating a breaking change.

@HarelM , good idea - I've added it as an additional build, avoiding any breakages at this time.

**Comment by snuup** [comment]
Would really need an ES6 module build.

**Comment by HarelM** [comment]
While the PR for this is ready (or close to it), this is a breaking change for Angular since Angular is using esm by default, so we'd need to wait for a breaking change version of maplibre, which is currently not planned as I don't think we have enough content right now for breaking changes. 
Nevertheless, you should be able to create this build locally from the PR if you need.

**Comment by birkskyum** [comment]
AMD / UMD / SystemJS is being deprecated in the typescript --module per TypeScript 6 (beta now) as well

https://devblogs.microsoft.com/typescript/announcing-typescript-6-0-beta/#deprecated:-amd-umd-and-systemjs-values-of-module

These old module format are starting to feel the heat.

**Comment by birkskyum** [comment]
> [!NOTE]
> When https://github.com/mapbox/mapbox-gl-rtl-text/releases eventually launches ESM support, or if we get impatient and make our own version that does, clean up `loadScript` helper in `src/source/worker.ts`

**Comment by hugobarragon** [comment]
@birkskyum hi, dont we need to export the package as esm spec, in modules and not in 1 file bundle, to allow esm to work properly and do for example tree shaking etc?

**Comment by HarelM** [comment]
@hugobarragon it's not clear what you are suggesting.
This PR was merged and the package released as a pre-release in npm.
If something is not working as expected do let us know, or better yet open a PR to fix it.
